src/schedule.rs

```diff
@@ -0,0 +1,190 @@
+            info!("Ignoring change on '{}' ({})", filename, path.display());
+            continue;
+        }
+
+        info!("Request to spawn a job");
+
+        // Check if another job is already in execution
+        let mut job = job_info.lock().unwrap();
+        if *job == JobInfo::Rustc {
+            info!("Another command is currently in execution: request denied");
+            continue;
+        } else if *job == JobInfo::User {
+            // Send kill signal. We can unwrap here, because `*job` is `Idle`
+            // until the channel is created. The result can be ignored: an
+            // error means that the other end has hung up. Since we hold the
+            // lock of `job`, this means that the other thread has panicked.
+            // The other end is deallocated when `job` is reset to `Idle`. The
+            // panicking case is handled below.
+            let _ = kill.unwrap().send(());
+
+            // After the kill signal was send, we have to wait for the thread
+            // to receive it and terminate itself. About unwrap: see above.
+            // We have to reset the job info ourselves, because the other
+            // thread is unable to, because we are holding the lock.
+            let res = thread.unwrap().join();
+            *job = JobInfo::Idle;
+            if res.is_err() {
+                info!("child thread panicked...")
+            }
+        }
+
+        // No other job is in execution: start new one.
+        *job = JobInfo::Rustc;
+
+        // Create channel to send kill signals
+        let (tx, rx) = channel();
+        kill = Some(tx);
+
+        let thread_commands = commands.clone();
+        let thread_job_info = job_info.clone();
+
+        thread = Some(
+            thread::spawn(move || {
+                execute_commands(&thread_commands, thread_job_info, rx)
+            })
+        );
+    }
+}
+
+/// Executes given commands in order (runs on extra thread)
+fn execute_commands(
+    commands: &[String],
+    job_info: Arc<Mutex<JobInfo>>,
+    kill: Receiver<()>,
+) {
+    // helper to update the "global" job information
+    let update_info = |new_info: JobInfo| {
+        // We can unwrap here: the only way `lock` returns an `Err` value is
+        // when the main thread has panicked. In this case this thread should
+        // panic, too.
+        let mut guard = job_info.lock().unwrap();
+        *guard = new_info;
+    };
+
+    debug!("Starting a command run!");
+
+    // Says if we received a kill signal
+    let mut abort = false;
+
+    // Iterate through all given commands and execute them in order
+    for command in commands {
+        if abort {
+            debug!("Aborting command run");
+            break;
+        }
+        let args: Vec<&str> = command.split_whitespace().collect();
+
+        println!("");
+        println!("$ cargo {}", command);
+
+        // Update global information about the running job
+        if args.get(0) == Some(&"run") {
```

Comment:
It's already split on whitespace, right? I figured it should be pretty safe. I don't see when this wouldn't work.